sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/BuilderAnnotationVisitor.java

```diff
@@ -136,70 +120,401 @@ public void visitClass(ClassElement element, VisitorContext context) {
-                if (writeMethod.isPresent()) {
-                    String propertyName = beanProperty.getSimpleName();
-                    TypeDef propertyTypeDef = TypeDef.of(beanProperty.getType());
-                    statementDefs.add(
-                        local.invoke(writeMethod.get(), thisVariable.field(propertyName, propertyTypeDef))
-                    );
-                }
+            createDefaultModifyPropertyMethod(classDefBuilder, beanProperty, returningExpressionProvider);
+        }
+    }
+
+    private static void createDefaultModifyPropertyMethod(ClassDef.ClassDefBuilder classDefBuilder,
+                                                          PropertyElement beanProperty,
+                                                          Function<VariableDef.This, StatementDef> returningExpressionProvider) {
+        TypeDef propertyTypeDef = TypeDef.of(beanProperty.getType());
+        FieldDef field = createField(beanProperty, propertyTypeDef);
+        classDefBuilder.addField(field);
+        String propertyName = beanProperty.getSimpleName();
+        classDefBuilder.addMethod(MethodDef.builder(propertyName)
+            .addModifiers(Modifier.PUBLIC)
+            .addParameter(propertyName, propertyTypeDef)
+            .build((self, parameterDefs) -> StatementDef.multi(
+                self.field(field).assign(parameterDefs.get(0)),
+                returningExpressionProvider.apply(self)
+            )));
+    }
+
+    private static FieldDef createField(PropertyElement beanProperty, TypeDef type) {
+        TypeDef fieldType = type.makeNullable();
+        if (!fieldType.isNullable()) {
+            throw new IllegalStateException("Could not make the field nullable");
+        }
+        FieldDef.FieldDefBuilder fieldDef = FieldDef.builder(beanProperty.getSimpleName())
+            .ofType(fieldType)
+            .addModifiers(Modifier.PROTECTED);
+        try {
+            beanProperty.stringValue(Bindable.class, "defaultValue").ifPresent(defaultValue ->
+                fieldDef.initializer(ExpressionDef.constant(beanProperty.getType(), fieldType, defaultValue))
+            );
+        } catch (IllegalArgumentException e) {
+            throw new IllegalStateException("Invalid or unsupported default value specified: " + beanProperty.stringValue(Bindable.class, "defaultValue").orElse(null));
+        }
+        return fieldDef.build();
+    }
+
+    private static void createSingularPropertyMethods(ClassDef.ClassDefBuilder classBuilder,
+                                                      PropertyElement beanProperty,
+                                                      Function<VariableDef.This, StatementDef> returningExpressionProvider) {
+        String propertyName = beanProperty.getSimpleName();
+        String singularName = beanProperty.stringValue(Singular.class).orElse(null);
+        if (singularName == null) {
+            if (propertyName.endsWith("s")) {
```

Comment:
not sure if we should make this logic smarter. Lombok seems to have stripped the logic from Rails to do this https://github.com/projectlombok/lombok/blob/master/src/core/lombok/core/handlers/singulars.txt

See https://github.com/projectlombok/lombok/blob/master/src/core/lombok/core/handlers/Singulars.java

Reply:
Added